File: www/cgi-bin/hello.pl
#!/usr/bin/perl
# Third CGI language (alongside Python and Bash) to demonstrate that CGI
# dispatch is driven purely by the configured file extension.
use strict;
use warnings;

my $method = $ENV{'REQUEST_METHOD'} || 'GET';
my $query  = $ENV{'QUERY_STRING'}   || '';
my $script = $ENV{'SCRIPT_NAME'}    || '';

my $body = '';
if ($method eq 'POST') {
    local $/ = undef;
    my $in = <STDIN>;          # read to EOF, as the CGI specification expects
    $body = defined($in) ? $in : '';
}

my $interpreter = $^X;
my $version     = $];
my $body_len    = length($body);

print "Content-Type: text/html\r\n";
print "\r\n";
print <<"HTML";
<!DOCTYPE html>
<html>
<head><title>Perl CGI</title></head>
<body style="font-family: sans-serif; background:#0f172a; color:#f8fafc; padding:40px;">
  <h2>Perl CGI is alive</h2>
  <p>Interpreter: $interpreter (Perl $version)</p>
  <p>SCRIPT_NAME: $script</p>
  <p>REQUEST_METHOD: $method</p>
  <p>QUERY_STRING: $query</p>
  <p>Body bytes read to EOF: $body_len</p>
  <a href="/" style="color:#38bdf8;">Return Home</a>
</body>
</html>
HTML

File: www/cgi-bin/hello.php
<?php
// Third CGI language, dispatched by the .php extension via php-cgi.
header("Content-Type: text/html");

$method = isset($_SERVER["REQUEST_METHOD"]) ? $_SERVER["REQUEST_METHOD"] : "";
$query  = isset($_SERVER["QUERY_STRING"]) ? $_SERVER["QUERY_STRING"] : "";
$body   = file_get_contents("php://input");

echo "<!DOCTYPE html>\n";
echo "<html><head><title>PHP CGI</title></head>\n";
echo "<body style=\"font-family: Arial, sans-serif;\">\n";
echo "<h1>Hello from PHP CGI</h1>\n";
echo "<p>REQUEST_METHOD: " . htmlspecialchars($method) . "</p>\n";
echo "<p>QUERY_STRING: " . htmlspecialchars($query) . "</p>\n";
echo "<p>Body length: " . strlen($body) . "</p>\n";
echo "<p>Working directory: " . htmlspecialchars(getcwd()) . "</p>\n";
echo "</body></html>\n";
